// @rozhkov/react-useful-hooks/dist/commonjs/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useChangeCounter = exports.useArgByRef = exports.useIsChanged = exports.useStateRef = exports.useMemoObject = exports.useMemoArray = exports.useStableCallback = exports.usePrevious = exports.useIsFirstRender = exports.useInit = void 0;
var useInit_1 = require("./useInit");
Object.defineProperty(exports, "useInit", { enumerable: true, get: function () { return __importDefault(useInit_1).default; } });
var useIsFirstRender_1 = require("./useIsFirstRender");
Object.defineProperty(exports, "useIsFirstRender", { enumerable: true, get: function () { return __importDefault(useIsFirstRender_1).default; } });
var usePrevious_1 = require("./usePrevious");
Object.defineProperty(exports, "usePrevious", { enumerable: true, get: function () { return __importDefault(usePrevious_1).default; } });
var useStableCallback_1 = require("./useStableCallback");
Object.defineProperty(exports, "useStableCallback", { enumerable: true, get: function () { return __importDefault(useStableCallback_1).default; } });
var useMemoArray_1 = require("./useMemoArray");
Object.defineProperty(exports, "useMemoArray", { enumerable: true, get: function () { return __importDefault(useMemoArray_1).default; } });
var useMemoObject_1 = require("./useMemoObject");
Object.defineProperty(exports, "useMemoObject", { enumerable: true, get: function () { return __importDefault(useMemoObject_1).default; } });
var useStateRef_1 = require("./useStateRef");
Object.defineProperty(exports, "useStateRef", { enumerable: true, get: function () { return __importDefault(useStateRef_1).default; } });
var useIsChanged_1 = require("./useIsChanged");
Object.defineProperty(exports, "useIsChanged", { enumerable: true, get: function () { return __importDefault(useIsChanged_1).default; } });
var useArgByRef_1 = require("./useArgByRef");
Object.defineProperty(exports, "useArgByRef", { enumerable: true, get: function () { return __importDefault(useArgByRef_1).default; } });
var useChangeCounter_1 = require("./useChangeCounter");
Object.defineProperty(exports, "useChangeCounter", { enumerable: true, get: function () { return __importDefault(useChangeCounter_1).default; } });
